### ISSUE TITLE
[BUG]CrystalAura multi thread bug 

### ISSUE THREAD

**Issue by SpartanB312** [WurstPlus/wurst-plus-three#266]
First is a logic bug but it will only happen on a very slow PC.
onUpdate() is invoked by TickEvent every tick. The delay between ticks is 50ms.So every 50ms the CrystalAura will launch the new thread to calculate. What if every calculation loop takes over 50ms?I think it would be terrible.

Secondly, after starting a calculation thread. It's wrong to get the static result immediately. Because the calculation is not completed in another thread. You can only get result calculated by last thread or null.

There is a example of concurrent multi thread AutoCrystal https://github.com/SexyTeam/Cursa/blob/main/src/main/java/club/eridani/cursa/module/modules/combat/CursaAura.java

**Comment by wallhacks0** [comment]
@Madmegsox1  multiplace calcs go brr

**Comment by Madmegsox1** [comment]
This is partly correct as the reason calcs are done on a different is not to speed the Auto Crystal up its to stop the Auto Crystal Lagging on bad computers, The fact that I am setting it to a static variable doesn't matter if it is null as we check if the EntityEnderCrystal is null, a bit later in the code here https://github.com/WurstPlus/wurst-plus-three/blob/d9108b1f5259bdeef4a80d502ca635c2dd61284d/src/main/java/me/travis/wurstplusthree/hack/hacks/combat/CrystalAura.java#L425

But going back to your first issue, the only way that it can take 50ms to run damage calculations is if the computer is a potato as the average CPU clock speed is > 3Ghz which is 3000000000  Calculations every second. But if it did take longer that 50ms then it would just return a null pointer which is handled.

**Comment by 0-x-2-2** [comment]
You are both retarded, also if your auto crystal is taking more than 1ms for 40 targets I am scared for your intellectual ability. 

I can tell you 30 reasons why multi-threading at all is a dangerous and bad idea especially with your disgustingly unsafe code and both of these implementations are absolutely horrible but creating a new thread at all is even worse as that will starve all other threads from CPU time.

**Comment by 0-x-2-2** [comment]
> This is partly correct as the reason calcs are done on a different is not to speed the Auto Crystal up its to stop the Auto Crystal Lagging on bad computers

Yeah, and it will be worse just for using another thread because your code and Minecraft are unsafe.

> the average CPU clock speed is > 3Ghz which is 3000000000 Calculations every second. But if it did take longer than 50ms then it would just return a null pointer which is handled.

This is not how CPUs work at all and the clock speed is not related to how many calculations are done at all and that does not include branching either as that is an entirely different story.

**Comment by wallhacks0** [comment]
nerd

**Comment by SpartanB312** [comment]
> You are both retarded, also if your auto crystal is taking more than 1ms for 10+ targets I am scared for your intellectual ability.
> 
> I can tell you 30 reasons why multi-threading at all is a dangerous and bad idea especially with your disgustingly unsafe code and both of these implementations are absolutely horrible but creating a new thread at all is even worse as that will starve all other threads from CPU time.

You are right.It's more important to optimize code than multi-threading in unsafe code.

**Comment by SpartanB312** [comment]
> This is partly correct as the reason calcs are done on a different is not to speed the Auto Crystal up its to stop the Auto Crystal Lagging on bad computers, The fact that I am setting it to a static variable doesn't matter if it is null as we check if the EntityEnderCrystal is null, a bit later in the code here
> 
> https://github.com/WurstPlus/wurst-plus-three/blob/d9108b1f5259bdeef4a80d502ca635c2dd61284d/src/main/java/me/travis/wurstplusthree/hack/hacks/combat/CrystalAura.java#L425
> 
> But going back to your first issue, the only way that it can take 50ms to run damage calculations is if the computer is a potato as the average CPU clock speed is > 3Ghz which is 3000000000 Calculations every second. But if it did take longer that 50ms then it would just return a null pointer which is handled.

It's really exist in some laggy computers bc it's a logical bug.Fix it or not ,it's up to you.

**Comment by SpartanB312** [comment]
> This is partly correct as the reason calcs are done on a different is not to speed the Auto Crystal up its to stop the Auto Crystal Lagging on bad computers, The fact that I am setting it to a static variable doesn't matter if it is null as we check if the EntityEnderCrystal is null, a bit later in the code here
> 
> https://github.com/WurstPlus/wurst-plus-three/blob/d9108b1f5259bdeef4a80d502ca635c2dd61284d/src/main/java/me/travis/wurstplusthree/hack/hacks/combat/CrystalAura.java#L425
> 
> But going back to your first issue, the only way that it can take 50ms to run damage calculations is if the computer is a potato as the average CPU clock speed is > 3Ghz which is 3000000000 Calculations every second. But if it did take longer that 50ms then it would just return a null pointer which is handled.

And I said you launch a thread every tick.But if each calculation took over 1 tick-length time.More and more thread will be launched

**Comment by SpartanB312** [comment]
> @Madmegsox1 multiplace calcs go brr

Multi-Threading is not multiplace bro

**Comment by SpartanB312** [comment]
> nerd

Cool boy confused Multi-Threading with MultiPlace : ) You reopened this issue to show how cool you are 👍

**Comment by Madmegsox1** [comment]
To be honest, I have been working on projects for a multi-national company. So I will come back to wurst + 3 and the community when I have the time to...

**Comment by hobrin** [comment]
@0-x-2-2 
> You are both retarded, also if your auto crystal is taking more than 1ms for 10+ targets I am scared for your intellectual ability.
> 
> I can tell you 30 reasons why multi-threading at all is a dangerous and bad idea especially with your disgustingly unsafe code and both of these implementations are absolutely horrible but creating a new thread at all is even worse as that will starve all other threads from CPU time.

You don't measure performance across computers and especially not across the internet in a time unit. I would be impressed if you can tell me 10 valid reasons of why you should be careful when multithreading. Creating a new thread does not starve all other threads from the CPU. The entire idea is that they can work simultaneously and in total if you sum up the amount of time taken to do a certain set of operations is way smaller.

If you are so against multi-threading then I wonder how you handle HTTP requests to servers on the main thread. Do you let your users have the feeling that their game is crashing when the entire game is just literally freezing for 2 seconds or so? That is not even mentioning if your internet is temporarily down because then the freeze can be up to 30 seconds long and by that point most unsuspecting users will have closed the game already.

And on top of that your github account should be deleted without question for using such language. I have never seen any programmer with such poor social skills like seem to have.

**Comment by CrystalPvpFan** [comment]
> To be honest, I have been working on projects for a multi-national company. So I will come back to wurst + 3 and the community when I have the time to...

pog

**Comment by cryrobtrwew** [comment]
This is a tragedy that "devs" are calling 3ghz means you can do 3000000000 calcs? this just shows how inexperienced the devs are now these days. I will prob agree with OX for once, he actually said smart shit about this.

**Comment by amongusenjoyer69** [comment]
You don't measure performance across computers and especially not across the internet in a time unit.

I would be impressed if you can tell me 10 valid reasons of why you should be careful when multithreading. Creating a new thread does not starve all other threads from the CPU. The entire idea is that they can work simultaneously and in total if you sum up the amount of time taken to do a certain set of operations is way smaller.

If you are so against multi-threading then I wonder how you handle HTTP requests to servers on the main thread. Do you let your users have the feeling that their game is crashing when the entire game is just literally freezing for 2 seconds or so? That is not even mentioning if your internet is temporarily down because then the freeze can be up to 30 seconds long and by that point most unsuspecting users will have closed the game already.

And on top of that your github account should be deleted without question for using such language. I have never seen any programmer with such poor social skills like seem to have.

**Comment by 0-x-2-2** [comment]
> I would be impressed if you can tell me 10 valid reasons of why you should be careful when multithreading.

Sounds like you are inexperienced if you can't see the thread-safety issues with this code with a game that is not thread-safe at all. If you are working with thread-safe code and multithreading is the job and not just a hammer to use on everything like most people like to think it is. If my hints are still not good enough think about this for a moment... what happens if you read from one thread and write to it from another at the same time to the CPU timer cycle? Thread safety is highly documented and the best practices are very public I am sure you are capable of reading into it on your own and come to the same conclusion that accessing memory between threads is not safe without making all code touching that memory safe.

> Creating a new thread does not starve all other threads from the CPU.

You can reproduce this yourself by creating the main thread then attempting to spin up 32 threads that just blocking on I/O so they should not even be using a measurable amount of CPU time, this will cause a noticeable reduction in ops while you are creating the threads for a few seconds or so and the solution in java is to just use a thread pool.

> If you are so against multi-threading then I wonder how you handle HTTP requests to servers on the main thread.

Handling I/O is not the same as executing code :) because one of them uses little or no CPU time.

> That is not even mentioning if your internet is temporarily down because then the freeze can be up to 30 seconds long and by that point most unsuspecting users will have closed the game already.

Sounds like this is offtopic as this is not a web service and is a feature in a Minecraft cheat client, but you can use my last answer.

> Do you let your users have the feeling that their game is crashing when the entire game is just literally freezing for 2 seconds or so?

All I/O should be done asynchronously, not sure where I denied this? 

> measure performance across computers 

No, you measure relative performance. If your computer can't run the game at around 20 FPS it also cannot process packets at the tick speed of the game due to how the game works so multithreading would be absolutely useless and when your frames start taking 50ms or longer adding 3ms won't change much at all and as I said if your shitty crystal placement algorithm is taking longer than a few ms you got some real problems.

> especially not across the internet in a time unit.

None of this data is even touching the internet and is processed locally it just happens that the result is used with the internet.

TLDR; Spinning up threads is not a hammer you should just use just because... you should use it because it is the proper tool for the job. In this situation, any multithreading is just wasting CPU cycles on absolutely nothing with the same result for up to 50ms.

**Comment by Madmegsox1** [comment]
Thanks @0-x-2-2 , you have really furthered my understanding of threads and thread safety. As I was blind to the fact that you it is important that you have this good coding practice for threads. 

Thanks again!

**Comment by Madmegsox1** [comment]
> This is a tragedy that "devs" are calling 3ghz means you can do 3000000000 calcs? this just shows how inexperienced the devs are now these days. I will prob agree with OX for once, he actually said smart shit about this.

@cryrobtrwew 
By this I am referencing the fact that the "3GHz" is the clock timing  / clock Speed, this is the number of cycles or instructions it can perform per second. The clock speed of a CPU is determined by its quartz-crystal circuit’s ability to provide a constant frequency output under varying load. During each clock cycle, the processor carries out or performs as many instructions as it can based on its instruction pipelining. Instruction pipelining or just pipelining is a series of instructions that a CPU can handle in parallel to maximize efficiency. The process of pipelining divides incoming instructions into a series of sequential steps that are then performed by the cores and threads of the CPU simultaneously or in parallel.

Sorry my comment can be misleading by saying that it can do 3000000000 cals per second, this number is based on of course its clock speed, but I doesn't take into account, load form the OS or other Software running on the application level and it doesn't take into account of threads and cores. So I would like to respectfully withdraw that comment.

**Comment by wallhacks0** [comment]
> You are both retarded, also if your auto crystal is taking more than 1ms for 10+ targets I am scared for your intellectual ability.
> 
> I can tell you 30 reasons why multi-threading at all is a dangerous and bad idea especially with your disgustingly unsafe code and both of these implementations are absolutely horrible but creating a new thread at all is even worse as that will starve all other threads from CPU time.

Your github account should be deleted without question for using such language. I have never seen any programmer with such poor social skills like seem to have.

**Comment by hobrin** [comment]
@0-x-2-2
I agree with the point about that you shouldn't start another thread in every situation. Previously I was assuming the idea that you were always against multithreading.

Your entire point about another thread 'starve all other threads from CPU time' is an exaggeration. The word starve implies that it kills the other threads however this is not the case, they will still run, however a tiny bit slower depending on how many cores you have on your CPU.

If you are crystal pvping, then I am assuming your main concern is the actual game's performance. Now there are 2 scenarios. Either you do the crystal placement calculations on the main thread, guarantying a delay of x over there.  Your other option is launching a new thread from the main thread taking y seconds, guaranteeing a way smaller delay, and the rest of the calculations will be handled simultaneously taking x amount of time.

> Sounds like you are inexperienced if you can't see the thread-safety issues with this code with a game that is not thread-safe at all.

You are avoiding the question.

> adding 3ms won't change much at all
That is true, if it only stays at adding 3ms.

Now I would like to get to the point, yes thread safety is something to put into consideration.

I am assuming here that your crystal aura algorithm only reads the player positions and does calculations based on that. Now reading a variable whilst it is being read or set elsewhere won't crash your game. The worst thing that can happen to my knowledge is that you get a newdated reading of a variable because it was being updated later in the main thread.

Now this might cause problems if let's say the x and y are old but the z is new.

I would like to end this by stating that multithreading is a tradeoff, between performance and thread safety.
